code/builders/builder-vite/src/plugins/webpack-stats-plugin.ts

```diff
@@ -0,0 +1,116 @@
+ */
+function stripQueryParams(filePath: string): string {
+  return filePath.split('?')[0];
+}
+
+/**
+ * We only care about user code, not node_modules, vite files, or (most) virtual files.
+ */
+function isUserCode(moduleName: string) {
+  return Boolean(
+    moduleName &&
+      !moduleName.startsWith('vite/') &&
+      !moduleName.startsWith('\x00') &&
+      !moduleName.startsWith('\u0000') &&
+      moduleName !== 'react/jsx-runtime' &&
+      !moduleName.match(/node_modules\//)
+  );
+}
+
+export type WebpackStatsPlugin = Plugin & { storybookGetStats: () => BuilderStats };
+
+export function pluginWebpackStats({ workingDir }: WebpackStatsPluginOptions): WebpackStatsPlugin {
+  /**
+   * Convert an absolute path name to a path relative to the vite root, with a starting `./`
+   */
+  function normalize(filename: string) {
+    // Do not try to resolve virtual files
+    if (filename.startsWith('/virtual:')) {
+      return filename;
+    }
+    // Otherwise, we need them in the format `./path/to/file.js`.
+    else {
+      const relativePath = path.relative(workingDir, stripQueryParams(filename));
+      // This seems hacky, got to be a better way to add a `./` to the start of a path.
+      return `./${relativePath}`;
+    }
+  }
+
+  /**
+   * Helper to create Reason objects out of a list of string paths
+   */
+  function createReasons(importers?: readonly string[]): Reason[] {
+    return (importers || []).map((i) => ({ moduleName: normalize(i) }));
+  }
+
+  /**
+   * Helper function to build a `Module` given a filename and list of files that import it
+   */
+  function createStatsMapModule(filename: string, importers?: readonly string[]): Module {
+    return {
+      id: filename,
+      name: filename,
+      reasons: createReasons(importers),
+    };
+  }
+
+  const statsMap = new Map<string, Module>();
+
+  return {
+    name: 'rollup-plugin-webpack-stats',
+    // We want this to run after the vite build plugins (https://vitejs.dev/guide/api-plugin.html#plugin-ordering)
+    enforce: 'post',
+    moduleParsed: function (mod) {
+      if (isUserCode(mod.id)) {
+        mod.importedIds
+          .concat(mod.dynamicallyImportedIds)
+          .filter((name) => isUserCode(name))
+          .forEach((depIdUnsafe) => {
+            const depId = normalize(depIdUnsafe);
+            if (statsMap.has(depId)) {
+              const m = statsMap.get(depId);
+              if (m) {
+                m.reasons = (m.reasons ?? [])
+                  .concat(createReasons([mod.id]))
+                  .filter((r) => r.moduleName !== depId);
+                statsMap.set(depId, m);
+              }
+            } else {
+              statsMap.set(depId, createStatsMapModule(depId, [mod.id]));
+            }
+          });
+      }
+    },
+
+    storybookGetStats() {
+      const stats = { modules: Array.from(statsMap.values()) };
+      return { ...stats, toJson: () => stats };
+    },
```

Comment:
I couldn't see a better way to do this, but I am not very familiar with Vite so there are probably much better patterns

Reply:
It looks to me like you could just keep the original [`generateBundle`](https://github.com/IanVS/vite-plugin-turbosnap/blob/main/src/index.ts#L116-L127) hook?

I'm not even sure how this works now, given that the the result here isn't written to any file based on what I can see?

Reply:
The original hook wrote directly to disk at the end of the build.

This instead provides a way for us to read the stats out at the end of the build process, and return it to `core-server` which will [write the stats file itself](https://github.com/storybookjs/storybook/blob/b629fa74293e70b9185242ec06580495291a09da/code/lib/core-server/src/build-static.ts#L186).